### PR TITLE
#283: Authenticator and Identifier in Result

### PR DESCRIPTION
**Pull request by Erwane**
attache Identifier and Authenticator to result.

This allow more control in Controller when there is multiple authenticators and should do actions when a specific authentication or identifier match.

Refs https://github.com/cakephp/authentication/issues/283

### CODE REVIEW

File: src/Authenticator/Result.php [hunk 1]
```diff
@@ -141,6 +141,7 @@ public function getIdentifier()
 
     /**
      * set the identifier who match the correct identity or null
+     * @param null|Authentication\Identifier\IdentifierInterface $identifier The matching identifier
```
Comment (dereuromark) [inline src/Authenticator/Result.php]: Please set |null to the end, fasely ones are not the primary types here.

File: src/Authenticator/ResultInterface.php [hunk 2]
```diff
@@ -79,9 +81,10 @@ public function getIdentifier();
 
     /**
      * set the identifier who match the correct identity or null
+     * @param null|Authentication\Identifier\IdentifierInterface $identifier The matching identifier
```
Comment (dereuromark) [inline src/Authenticator/ResultInterface.php]: Invalid FQCN, \ missing in front of the type

File: src/Authenticator/Result.php [hunk 3]
```diff
@@ -45,15 +46,29 @@ class Result implements ResultInterface
      */
     protected $_errors = [];
 
+    /**
+     * successful identifier or null
```
Comment (dereuromark) [inline src/Authenticator/Result.php]: There seems to be a newline missing

File: src/Authenticator/Result.php [hunk 4]
```diff
@@ -113,4 +131,42 @@ public function getErrors()
     {
         return $this->_errors;
     }
+
+    /**
+     * return null or the identifier who match the correct identity
```
Comment (dereuromark) [inline src/Authenticator/Result.php]: Same here.
Also please real sentences. Capital letter start, full stop at end.

**Comment by dereuromark** [comment]
Looks good.

File: src/Authenticator/Result.php (hunk 3 above)
Comment (Erwane) [reply src/Authenticator/Result.php]: i don't understand why.

```
$ composer cs-check
> phpcs --colors -p src/ tests/
............................................................ 60 / 91 (66%)
...............................                              91 / 91 (100%)


Time: 2.68 secs; Memory: 12MB
```

Comment (dereuromark) [reply src/Authenticator/Result.php]: Some of those are not properly sniffed yet, missing sniffer rules.
But you can see the other doc lines on how this is handled so far.

**Comment by Erwane** [comment]
Why phpstan fail when my unit tests are good ?

**Comment by dereuromark** [comment]
The interface doesnt have getSuccessfulIdentifier() does it?
You can add it there as annotation above, this way we keep BC here.

**Comment by dereuromark** [comment]
Is the target here 1.next or master? 1.next looks super outdated.

**Comment by Erwane** [comment]
i've took 1.next because master looks for cake4 no ?

**Comment by dereuromark** [comment]
No, I think 2.x is for cake4.
```
/**
 * @method type xyz()
 */
```
can be used to please phpstan. If this is a good idea or not, I cannot answer at this point.
Maybe wait for @ADmad etc to weigh in here.

**Comment by Erwane** [comment]
ok, close this PR please, i will make a new one to master.
thanks a lot.

**Comment by dereuromark** [comment]
You dont have to close, you can easily just set a different target as well, if you would like to.

**Comment by dereuromark** [comment]
Now https://github.com/cakephp/authentication/pull/298

**Pull request closed without merging**